src/bt-core-io.c

```diff
@@ -465,27 +468,78 @@ energy_info_cb(bt_activity_energy_info *energy_info ATTRIBS(UNUSED))
   /* nothing to do */
 }
 
+struct alarm_state {
+  alarm_cb cb;
+  void* data;
+};
+
+static enum ioresult
+alarm_event_in(int fd, void* data)
+{
+  uint64_t num_expirations;
+  ssize_t res;
+
+  res = TEMP_FAILURE_RETRY(read(fd, &num_expirations, sizeof(num_expirations)));
+  if (res < 0) {
+    ALOGE_ERRNO("read");
+    return IO_ABORT;
+  }
+
+  remove_timer(fd);
+  free(data);
+  return IO_OK;
+}
+
+static enum ioresult
+alarm_event(int fd, uint32_t events, void* data)
+{
+  enum ioresult res;
+
+  if (events & EPOLLIN) {
+    res = alarm_event_in(fd, data);
+  } else {
+    ALOGW("unsupported event mask: %u", events);
+    res = IO_OK;
+  }
+  return res;
+}
+
 static bool
-set_wake_alarm(uint64_t delay_millis ATTRIBS(UNUSED),
-               bool should_wake ATTRIBS(UNUSED),
-               alarm_cb cb ATTRIBS(UNUSED),
-               void *data ATTRIBS(UNUSED))
+set_wake_alarm_os_callout(uint64_t delay_millis, bool should_wake,
+                          alarm_cb cb, void* data)
 {
-  /* FIXME: need to be implemented in later patches */
+  struct alarm_state* alarm_state;
+  int clockid;
+
+  alarm_state = malloc(sizeof(*alarm_state));
+  if (!alarm_state)
+    return false;
+
+  alarm_state->cb = cb;
+  alarm_state->data = data;
+
+  clockid = should_wake ? CLOCK_REALTIME_ALARM : CLOCK_REALTIME;
```

Comment:
CLOCK_REALTIME is definitely incorrect. The name 'realtime' is misleading, because these clocks count wall-clock time. If someone modifies the time or date setting on th device, CLOCK_REALTIME_* will mirror this change and the timeout will become incorrect.

I guess 'should_wake' means 'wakeup from suspend'? Google wrote Android Alarm timers for this use case. Linux contains this feature as extension to timerfds. The correct clocks should be CLOCK_BOOTTIME and CLOCK_BOOTTIME_ALARM. A quick grep through the Flame's kernel sources suggest that it's available to us.

Lwn.net has an article on the topic at [1].

I'll r- for now, because this is a significant problem.

[1] http://lwn.net/Articles/429925/